### ISSUE TITLE
Pulumi workspace provisioner

### ISSUE THREAD

**Issue by hh** [coder/coder#4528]
Conversation starter around having a second provisioner based on pulumi.

**Comment by hh** [comment]
I'd like to understand a bit more about the provisioner interface and what's required from it.

**Comment by bpmct** [comment]
I think it's unlikely for us to develop another provisioner before a major release (v2.0) which plan to start work on at the end of the month.

It'd be helpful to know your use case with Pulumi and where Terraform falls short.

Perhaps we could also set up a sync to go over how our provisioner API works and how potentially could be extended if you're open to forking/submitting a PR?

**Comment by bpmct** [comment]
I think it's unlikely for us to develop another provisioner before a major release (v2.0) which plan to start work on at the end of the month.

It'd be helpful to know your use case with Pulumi and where Terraform falls short.

Perhaps we could also set up a sync to go over how our provisioner API works and how potentially could be extended if you're open to forking/submitting a PR?

**Comment by bpmct** [comment]
I think it's unlikely for us to develop another provisioner before a major release (v2.0) which we plan to start work on at the end of the month. ("It'll be a while before")

It'd be helpful to know your use case with Pulumi and where Terraform falls short.

Perhaps we could also set up a sync to go over how our provisioner API works and how potentially could be extended if you're open to forking/submitting a PR?

**Comment by bpmct** [comment]
I think it's unlikely for us to develop another provisioner before a major release (v2.0) which we plan to start work on at the end of the month. ("It'll be a while before")

It'd be helpful to know your use case with Pulumi and where Terraform falls short.

Perhaps we could also set up a sync to go over how our provisioner API works and how potentially could be extended if you're open to forking/submitting a PR?

**Comment by ammario** [comment]
We're considering making the provisioner layer extensible so people can create/bring third-party provisioners that Coder doesn't have to maintain.

**Comment by usrbinkat** [comment]
+1 on Pulumi provider being a valuable effort.

TF can be powerful, but as a DSL it doesn't expose the full native logic available with traditional programming languages, and it's features demand yet-another learning curve. Pulumi brings native programming language expressions to the table as a first-class feature.

**Comment by hh** [comment]
The provisionersdk proto is here isn't terribly complex:
https://github.com/coder/coder/blob/main/provisionersdk/proto/provisioner.proto

And a quick brain dump from a side-convo with @deansheather :

> there are some parameters sent to the provisioner
> currently the terraform provisioner parses the state and converts it into our GRPC types and ships it off to coder
> a new provider would need coder metadata types to send to coder
> the stuff send back on complete is mainly a list of Resource (name, type, and agent details if any)

**Comment by deansheather** [comment]
Reopening as it's related to a recent ticket.

**Comment by dirien** [comment]
_Disclaimer: I work at Pulumi._

A customer of us, told me they evaluate Coder for their development environments! So would be nice to have Pulumi support for provisioning too! 

Happy to help on this if needed or connect with our engineers here at Pulumi.

**Comment by dirien** [comment]
Disclaimer: I work at Pulumi.

Would be nice to have Pulumi support for provisioning too!

Happy to help on this if needed.

**Comment by dirien** [comment]
> Disclaimer: I work at Pulumi.
> 
> 
> 
> Would be nice to have Pulumi support for provisioning too!
> 
> 
> 
> Happy to help on this if needed.

Still happy to contribute/help if wanted!

**Comment by ammario** [comment]
@dirien really appreciate your offer to help here. We probably want to take a second serious look at the provisioner API before adding an additional type. Are you seeing user demand on Pulumi's end? I'd love to sync live if you're open to it, if so, please reach out to me at ammar@coder.com.

**Comment by ammario** [comment]
For the record, we're still interested in doing this but would like to work more closely with the Pulumi team on it. If anyone from there is interested, please reach out to me at ammar@ code.com.

**Comment by hh** [comment]
Glad to see new eyes on this : 👀

**Comment by bpmct** [comment]
An update here: We recently added support for external workspaces, isn't exactly an additional provisioner but a way to attach externally provisioned workspaces to Coder! 

It's currently licensed as Premium, but if there is a strong community use case feel free to reach out. 

You can learn about external workspaces here: https://coder.com/docs/admin/templates/managing-templates/external-workspaces

**Comment by michaelbrewer** [comment]
The reason to support other provisioners would be performance and flexibility. 

Currently terraform adds 6-16 seconds delay to getting any workspace. Even pre-builds. 

Ideally the enter end to end flow would be a couple seconds. Making it feel nearly instant. Currently end to end can be 16 seconds and up to a minute when things are not bundled in the workspace.

**Comment by bpmct** [comment]
Yeah. We will be working on near-instant workspaces soon, where workspaces can be provisioned and/or assigned without a TF operation

**Comment by hh** [comment]
Keen on this! ^^
